test/cache/map_cache_test.dart

```diff
@@ -66,20 +67,39 @@ main() {
     });
 
     test("should not make multiple requests for the same key", () async {
+      final completer = Completer<String>();
       int count = 0;
 
       Future<String> loader(String key) {
         count += 1;
-        return new Future.delayed(
-            const Duration(milliseconds: 1), () => "test");
+        return completer.future;
       }
 
-      await Future.wait([
+      final futures = Future.wait([
         cache.get("test", ifAbsent: loader),
         cache.get("test", ifAbsent: loader),
       ]);
 
+      completer.complete('bar');
       expect(count, equals(1));
+      expect(await futures, equals(['bar', 'bar']));
+    });
+
+    test("should not cache a failed request", () async {
+      int count = 0;
+
+      Future<String> loader(String key) async {
+        count += 1;
+        throw new StateError("Request failed");
+      }
+
+      await expectLater(
+          await () => cache.get("test", ifAbsent: loader), throwsStateError);
+      await expectLater(
+          await () => cache.get("test", ifAbsent: loader), throwsStateError);
+
+      expect(count, equals(2));
+      expect(await cache.get('test'), isNull);
```

Comment:
Could you also follow-up by a third attempt that succeeds and verify that the cache successfully returns a value?

Reply:
Done